### DISCUSSION TITLE
Inquiry on the differences between two ways of calculating the dimensional mean

### DISCUSSION THREAD

**Discussion by ChenyaoYang123** [pydata/xarray#6081] [Q&A] [answered]
Hello,
I am new to Xarray, and I am having a simple question.
Let´s see we have a 3-D xarray _da_ with dimension (time, lat, lon). 
So what is the difference between `da.mean(dim="lat",skipna=True).mean(dim="lon",skipna=True)` and `da.mean(dim=["lon","lat"], skipna=True)`? The data I have show clearly a difference between the two ways of calculating mean.
Many thanks in advance for your kind attentions and help.

Best regards
Chenyao

**Comment by andersy005** [comment, marked as answer]
The mean operation is the sum of the values divided by the number of values (by definition). Unfortunately, adding all values up and dividing accumulates floating-point errors. So, presuming your `da` consists of floating-point numbers, the differences you are seeing between 

```python
da.mean(dim="lat",skipna=True).mean(dim="lon",skipna=True)
```

and 

```python
da.mean(dim=["lon","lat"], skipna=True)
```

are due to differences in accumulated floating-point errors in both operations. To minimize the accumulated floating error, you may want use `da.mean(dim=["lon","lat"], ...)` instead of the chained `.mean("lon", ....).mean("lat", ....)` operation.

**Reply by max-sixty** [reply]
The other case where there can be differences is where there are nulls:

```python

In [6]: xr.DataArray([[10, 1], [np.nan, 5]]).mean("dim_0", skipna=True).mean("dim_1", skipna=True)
Out[6]:
<xarray.DataArray ()>
array(6.5)

In [7]: xr.DataArray([[10, 1], [np.nan, 5]]).mean(skipna=True)
Out[7]:
<xarray.DataArray ()>
array(5.33333333)
```

**Comment by ChenyaoYang123** [comment]
Thank you both for your kind help. In my case, NaN values are quite common in the dataset so I think it will essentially explain the difference in the computed mean values. Besides, I also think the floating-point issue is really interesting and definitely worth noting.
